clang/lib/CodeGen/SYCLLowerIR/LowerWGScope.cpp

```diff
@@ -921,7 +921,7 @@ Value *spirv::genLinearLocalID(Instruction &Before, const Triple &TT) {
       unsigned Align = M.getDataLayout().getPreferredAlignment(G);
       G->setAlignment(Align);
     }
-    Value *Res = new LoadInst(G, "", &Before);
+    Value *Res = new LoadInst(G, "", true, &Before);
```

Comment:
This is the wrong way to fix the problem.
Please, add "convergent" attribute to the barrier function at line 944.
NOTE: it's not the only function missing this attribute.

Please, add a unit test for this pass to clang LIT test suite.

Reply:
@Naghasan, do we need add this attribute to `_Z22__spirv_ControlBarrierN5__spv5ScopeES0_j` implementation in libclc?

Reply:
That's probably safer to do so, regardless of how llvm handles attribute merging.

Reply:
Thx, I am going to try add attribute but I am not sure that this will fix the problem.
Could you please take a look at the example I provided in the issue: https://github.com/intel/llvm/issues/1258

Definition contains convergent attribute:
```
; Function Attrs: convergent nounwind
declare void @llvm.nvvm.barrier0() #3

```
Nevertheless illegal transformation happens.

Reply:
```llvm
  tail call void @llvm.nvvm.barrier0() #4
...
attributes #3 = { convergent nounwind }
attributes #4 = { nounwind }
```
Call must be convergent too i.e. `#3`.
Note these might require it too:
```llvm

; Function Attrs: nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.tid.x() #2

; Function Attrs: nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.tid.y() #2

; Function Attrs: nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.tid.z() #2

; Function Attrs: nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ntid.y() #2

; Function Attrs: nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ntid.z() #2

; Function Attrs: nounwind readnone
declare i32 @llvm.nvvm.read.ptx.sreg.ctaid.x() #2
```

Reply:
I'm almost 100% sure that applying `convergent` will fix the problem.
I'm 146% sure that using `volatile` for load/store instructions is not the right approach.

Reply:
Discussed with Alexey offline, unfortunately convergent attribute doesn't help. As I described the problem, it happens because of globals AA, convergent attribute doesn't affect alias analysis. It looks like problem is how global variables are created in LowerWGScope pass. They are created with internal linkage. As a result alias analysis thinks that this global doesn't escape and gives wrong answer that barrier references but doesn't modify memory location of the global.   It looks like globals variables produced by LowerWGScope should have external linkage. I am working on the new patch.

Reply:
Created separate PR with new patch https://github.com/intel/llvm/pull/1257. For some reasons, I am not able to add Victor as a reviewer. This PR is going to be closed.